feat(ImageUpdater): Add ImageUpdater controller

dkarpele opened the pull request:
**What type of PR is this?**

/kind enhancement

**What does this PR do / why we need it**:
Image Updater: Provide option in Argo CD CR to enable/disable image updater

**Have you updated the necessary documentation?**

* [ ] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes #?

**How to test changes / Special notes to the reviewer**:

svghadi commented on controllers/argocd/imageUpdater.go:
```diff
@@ -0,0 +1,730 @@
+	podSpec.Containers = []corev1.Container{{
+		Command:         []string{"/manager"},
+		Args:            []string{"run"},
+		Image:           argoutil.CombineImageTag(DefaultImageUpdaterImage, DefaultImageUpdaterTag),
```
We should make this configurable via env var so that downstream images can be set through the CSV in gitops-operator. 

For ref: https://github.com/argoproj-labs/argocd-operator/blob/ef7197954c1d2ba5b3f733406d294956a78d423a/common/keys.go#L176

svghadi commented on controllers/argocd/imageUpdater.go:
```diff
@@ -0,0 +1,730 @@
+		Command:         []string{"/manager"},
+		Args:            []string{"run"},
+		Image:           argoutil.CombineImageTag(DefaultImageUpdaterImage, DefaultImageUpdaterTag),
+		ImagePullPolicy: corev1.PullAlways,
```
There are ongoing efforts to make the pull policy configurable, just wanted to highlight that.

svghadi reviewed (changes requested):
Thanks @dkarpele for working on this. Looks good overall. I have left some comments. 

- We should also bundle the ImageUpdater CRD into operator, similar to how we handle others [here](https://github.com/argoproj-labs/argocd-operator/tree/master/config/crd/bases).

- In the PR description, it mentions pulling from the crd branch in go.mod:
   ```
   go get github.com/argoproj-labs/argocd-image-updater@crd
   ```
   Is there a specific reason for using the crd branch instead of main or a release branch/tag?

dkarpele commented on controllers/argocd/imageUpdater.go:
```diff
@@ -0,0 +1,730 @@
+
+func (r *ReconcileArgoCD) reconcileImageUpdaterConfigurationCR(cr *argoproj.ArgoCD) error {
+
+	defaultImageUpdaterConfigurationCR := &imageupdaterv1alpha1.ImageUpdater{
```
Users can modify this CR or create a new one.
My initial idea was to create a basic ImageUpdater CR that can help user to start working with image updater. @chengfang do you think we should have default ImageUpdater CR or not?

chengfang replied:
It should be straightforward for users to create their ImageUpdater CR, and likely in their application git repo. I imagine most users will probably not actually use or modify this default CR for their purpose. I feel we can remove it to avoid the runtime overhead.

dkarpele commented:
> * In the PR description, it mentions pulling from the crd branch in go.mod:
>   ```
>   go get github.com/argoproj-labs/argocd-image-updater@crd
>   ```
  
>   Is there a specific reason for using the crd branch instead of main or a release branch/tag?

The reason is that at the moment crd development of image updater goes in crd branch (not master). When we make the crd version "main" we will update the dependencies here to use
```
go get github.com/argoproj-labs/argocd-image-updater@main
```

dkarpele commented on Makefile:
```diff
@@ -101,6 +101,10 @@ fmt: ## Run go fmt against code.
+get-image-updater-crd: ## Download Image Updater CRD.
+	@echo "downloading image updater crd"
+	@curl -sSLo config/crd/bases/argocd-image-updater.argoproj.io_imageupdaters.yaml https://raw.githubusercontent.com/argoproj-labs/argocd-image-updater/crd/config/crd/bases/argocd-image-updater.argoproj.io_imageupdaters.yaml
+
```
I was thinking it could be useful in case someone wants to update `argocd-image-updater.argoproj.io_imageupdaters.yaml`. The original of this file is located in Image Updater repo and to reflect the future changes in this crd the target `get-image-updater-crd:` can be used.

dkarpele commented:
@svghadi 
- I removed ImageUpdater CR as discussed.
- As a result we don't need to updater go.mod with ImageUpdater dependency.
- I added unit tests in `controllers/argocd/image_updater_test.go`.
- e2e tests will be created later (end of October), after ImageUpdater CRD release 1.0.0. This is because we can't create Deployment properly (image link leads to an old version of ImageUpdater), see GITOPS-8016.
